java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,155 @@
+      .filter(memberSymbol -> !memberSymbol.usages().isEmpty())
+      .filter(memberSymbol -> !isUsedOutsideMethods(memberSymbol, classSymbol))
+      .forEach(this::checkPrivateField);
+
+    fieldsReadOnAnotherInstance.clear();
+    fieldsLiveInMethodEntryPerMethod.clear();
+  }
+
+  private static boolean isPrivateField(Symbol memberSymbol) {
+    return memberSymbol.isPrivate() && memberSymbol.isVariableSymbol();
+  }
+
+  private static boolean isUsedOutsideMethods(Symbol privateFieldSymbol, TypeSymbol classSymbol) {
+    for (IdentifierTree usageIdentifier : privateFieldSymbol.usages()) {
+      Tree containingClassOrMethod = containingClassOrMethod(usageIdentifier);
+
+      if (containingClassOrMethod.is(Kind.CLASS)
+        || (containingClassOrMethod.is(Kind.METHOD) && !((MethodTree) containingClassOrMethod).symbol().owner().equals(classSymbol))) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private void checkPrivateField(Symbol privateFieldSymbol) {
+    if (fieldsReadOnAnotherInstance.contains(privateFieldSymbol)) {
+      return;
+    }
+
+    for (Set<Symbol> fieldsLiveInMethodEntry : fieldsLiveInMethodEntryPerMethod.values()) {
+      if (fieldsLiveInMethodEntry.contains(privateFieldSymbol)) {
+        return;
+      }
+    }
+
+    reportIssue(((VariableTree) privateFieldSymbol.declaration()).simpleName(), String.format(MESSAGE, privateFieldSymbol.name()));
+  }
+
+  private static Tree containingClassOrMethod(IdentifierTree usageIdentifier) {
+    Tree parent = usageIdentifier;
+    do {
+      parent = parent.parent();
+    } while (!parent.is(Kind.METHOD, Kind.CLASS));
+
+    return parent;
+  }
+
+  private void initFieldsMetaData(TypeSymbol classSymbol) {
+    fieldsReadOnAnotherInstance = FieldsReadOnAnotherInstanceVisitor.getFrom(classSymbol);
+
+    for (Symbol memberSymbol : classSymbol.memberSymbols()) {
+
+      if (memberSymbol.isMethodSymbol() && memberSymbol.declaration() != null) {
+        MethodTree methodTree = (MethodTree) memberSymbol.declaration();
+
+        if (methodTree.block() != null) {
+          CFG cfg = CFG.build(methodTree);
+          LiveVariables liveVariables = LiveVariables.analyze(cfg, true);
+          fieldsLiveInMethodEntryPerMethod.put(methodTree, liveVariables.getIn(cfg.entry()));
+        }
+      }
+    }
+  }
+
+  private static class FieldsReadOnAnotherInstanceVisitor extends BaseTreeVisitor {
+
+    private Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+
+    static Set<Symbol> getFrom(TypeSymbol classSymbol) {
+      FieldsReadOnAnotherInstanceVisitor fieldsReadOnAnotherInstanceVisitor = new FieldsReadOnAnotherInstanceVisitor();
+      fieldsReadOnAnotherInstanceVisitor.scan(classSymbol.declaration());
+      return fieldsReadOnAnotherInstanceVisitor.fieldsReadOnAnotherInstance;
+    }
+
+    @Override
+    public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+      if (tree.expression().is(Kind.IDENTIFIER)) {
```

Comment:
This will prevent you to get a lot of fields that could be read from another instance : 

```
class A {
  int field;
   A foo() {}
   void test() {
           new A().field; // read and MSE expression is not an identifier
           A a = new A();
           a.foo().field; // read and MSE expression is not an identifier
   }
}
```